lib/bolt/pal/yaml_plan.rb

```diff
@@ -103,6 +139,66 @@ def return_type
         Puppet::Pops::Types::TypeParser.singleton.parse('Boltlib::PlanResult')
       end
 
+      def validate_step_keys(step_keys)
+        action = step_keys.intersection(STEP_KEYS.keys.to_set).to_a
+        unless action.count == 1
+          if action.count > 1
+            raise Bolt::Error.new("Multiple action keys specified in a single step '#{action.join(',')}'",
+                                  "bolt/invalid-plan")
+          else
+            raise Bolt::Error.new("No action specified for step", "bolt/invalid-plan")
+          end
+        end
+
+        # For validated step action, ensure only valid keys
+        unless STEP_KEYS[action.first]['allowed_keys'].superset?(step_keys)
+          illegal_keys = step_keys - STEP_KEYS[action.first]['allowed_keys']
+          raise Bolt::Error.new("The '#{action.first}' step does not support: '#{illegal_keys.to_a.join(',')}' key(s)",
```

Comment:
I need to figure out consistent error reporting. Specifically how we point to *where* the error is found.  
The first thing that popped in to my mind was identifying steps by number (because steps are not required to have `name`).
It would be most useful as a user if we could point to a line number instead of a step number/name, but I do not see how that will be preserved with existing yaml parsing.

Reply:
Preserving line numbers is not worth it just for reporting validation. We can revisit this if there is demand. For now we will *always* reference step number and *attempt* to reference step name.